04-onion-routing.md

```diff
@@ -1068,6 +1178,21 @@ key, and computes the HMAC, using each hop's `um` key.
 The origin node can detect the sender of the error message by matching the
 `hmac` field with the computed HMAC.
 
+For trampoline payments, the flow is the same. Intermediate trampoline nodes
+first decrypt the downstream error using the `ammag` and `um` keys for their
+forward path. If the error comes from an intermediate node _before_ the next
+trampoline node, they may replace it with their own error for the origin node,
+otherwise they must encrypt on top of the next trampoline node's error.
```

Comment:
Encrypt on top for attributable failures would I think mean that they'd just process it as usual - shifting hmacs and adding their own?

I am not sure what the intermediate nodes between the two trampolines would do though. Would that require two `attribution_data` blocks?

Reply:
I've been reading up on trampoline more. It indeed seems that for attributable failures, you'd have two blocks of attribution data. The inner block can possibly be stored together with the failure message?

Reply:
How we implemented this in LDK is to just drop the attribution data if it cannot be interpreted by the sender.

Reply:
I don't think you need a second block of attribution data, there is only one failure packet and there should be only one attribution data to go with it.
The process for the attribution data is the same as for the failure packet:
- If the error comes from an intermediate node that you've chosen and not from a trampoline node: You should maybe retry and if it doesn't succeed, create a failure yourself, you are the failing trampoline node.
- If the error comes from a trampoline node: You unwrap the failure packet and the attribution data until you reach the next trampoline node and then wrap it with both shared secrets (the one from the sender and the one from the previous trampoline node).

Reply:
The second case indeed, trampoline node A receiving an error from a next trampoline node B. The attribution data that goes along with that isn't interpretable by the sender, so I think there is nothing to do other than drop it? And then let trampoline node A instantiate a new attr data block for the sender, to at least get attribution for the remaining downstream nodes. Not 100% sure if we are talking about the same thing, because you mentioned 'wrap it with both shared secrets'. You can't do that with attribution data?

Reply:
"wrap it with both shared secrets" it what the spec requires for the failure packet and we can do the same to the attribution data for consistency. Why do you think we can't do that? It's just xoring with a pseudo-random stream.

Reply:
If it makes it easier, you can just add a virtual hop from the trampoline node to itself. The trampoline node would populate the attribution data twice, once with each shared secret.

Reply:
We can xor the attribution data for sure, but unlike the failure message, which is ultimately readable by the sender, attribution data produced by 'inner route' nodes isn't. That attribution data is useless to begin with (to the sender), and no point in trying to wrap it so that the sender can read it. Unless I am missing something?

Reply:
Ah got it now, thanks. My thought was that once these inner route nodes start processing the attribution data, it becomes unusable. But as you point out, the trampoline node just shifts that the other way again to 'recover' the attribution data as it was passed back by the next trampoline node.